Reorg: Fix build artifacts action

brbrr opened the pull request:
Allows Jetpack to build again.

#### Changes proposed in this Pull Request:
*

#### Jetpack product discussion
1199714989172614-as-1199714989172639

#### Does this pull request change what data or activity we track or use?
n/a

#### Testing instructions:
* Check "artifacts" and "JS Tests" actions. Make sure they are working fine. 
* Launch a JN site for this branch.
* `cd projects/plugins/jetpack && yarn build`

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* n/a

brbrr committed:
Update Jetpack path

brbrr committed:
Update Jetpack path and move svnignore

brbrr committed:
add some debugging output

brbrr committed:
update artifact path

brbrr committed:
:shrug:

brbrr committed:
Update path to root eslint config

brbrr committed:
update js test action and scripts

brbrr committed:
use working-direcotry

brbrr committed:
Move eslint and jest configs into jetpack directory

anomiex commented on .eslintignore:
```diff
@@ -1,24 +0,0 @@
-/_inc/client/**/test/*.js
-/_inc/jetpack-modules.models.js
-/_inc/postmessage.js
-/docker/wordpress-develop/
-/docker/wordpress/
```
We should keep these around somewhere, whether in a root-level .eslintignore or in tools/docker

anomiex commented on .eslintignore:
```diff
@@ -1,24 +0,0 @@
-/packages/*/wordpress
-
-/modules/widgets/search/js/search-widget-admin.js
-/vendor/
```
This too.

anomiex commented on package.json:
```diff
@@ -65,7 +64,10 @@
 		"docker:update-core-unit-tests": "yarn docker:compose exec wordpress svn up /tmp/wordpress-develop/tests/phpunit/data/ /tmp/wordpress-develop/tests/phpunit/includes",
 		"docker:wp": "yarn docker:compose exec wordpress wp --allow-root --path=/var/www/html/",
 		"install-if-deps-outdated": "yarn check 2> /dev/null || yarn install --check-files --production=false --frozen-lockfile",
-		"lint": "eslint --ext .js,.jsx . _inc extensions modules",
```
We should keep this. We'll still want to lint js in tools, packages, and so on.

anomiex commented on package.json:
```diff
@@ -65,7 +64,10 @@
-		"lint": "eslint --ext .js,.jsx . _inc extensions modules",
+		"jetpack:build": "yarn jetpack:cd && yarn build-concurrently",
+		"jetpack:cd": "cd projects/plugins/jetpack",
+		"jetpack:lint": "yarn jetpack:cd && yarn lint",
```
`jetpack:lint` is probably not needed.

We may not even need a `yarn lint` in Jetpack's package.json, since we should be linting at the monorepo level. I dropped all those kinds of scripts from Jetpack's composer.json already (which reminds me, I forgot to remove the references to them from Jetpack's package.json...).

anomiex commented on projects/plugins/jetpack/.eslintrc.js:
```diff
@@ -1,3 +1,10 @@
+// module.exports = {
+// 	// Use root level ESlint configuration.
+// 	// JavaScript files inside this folder are meant to be transpiled by Webpack.
+// 	root: true,
+// 	extends: [ '../../../.eslintrc.js' ],
+// };
+
```
???

anomiex commented on .eslintignore:
```diff
@@ -1,24 +0,0 @@
-*.min.js
```
We currently have some .min.js files in projects/packages/lazy-images/ that were being ignored by this, but aren't anymore because you moved the file. We probably want a projects/packages/lazy-images/.eslintignore to re-ignore them.

brbrr replied:
moved file back to root

anomiex reviewed (commented):
In addition to the inline comments, this needs a merge of feature/reorg and then yarn.lock needs updating.

brbrr committed:
Merge remote-tracking branch 'origin/feature/reorg' into fix/build-artifacts-action

brbrr committed:
move eslintignore back to the root

brbrr committed:
update eslintignore

brbrr committed:
move eslint config back to repo root

brbrr committed:
update root node dependencies

brbrr committed:
convert indentation to Tabs in itersectionobserver-ployfill

brbrr committed:
run lint action from root directory

brbrr committed:
remove jetpack-specific scripts from root  package.json

brbrr committed:
add missing `commander` dependency

brbrr reviewed (commented):
I moved linting-related files back to the root. 

I'd like to note that `eslint` configuration relies on root `node_modules`, which means we need to `yarn install` on root to be able to run linting (and pre-commit scripts as well). Also, `eslint` does not support multiple `.eslintignore` files along the path, which potentially would lead to huge paths in root's `.eslintignore`. Alternatively, we could take advantage of `ignorePatterns` in config file(s)

brbrr committed:
Update path and add doc bock in webpack config

anomiex commented:
> Also, `eslint` does not support multiple `.eslintignore` files along the path, which potentially would lead to huge paths in root's `.eslintignore`. Alternatively, we could take advantage of `ignorePatterns` in config file(s)

Yeah, I had thought it would because it does allow for `.eslintrc` along the path, but apparently it doesn't.`

brbrr committed:
Update .eslintignore

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

anomiex commented on projects/packages/lazy-images/src/js/intersectionobserver-polyfill.js:
```diff
@@ -23,17 +23,17 @@
 		'IntersectionObserverEntry' in window &&
```
Now that you fixed .eslintignore, you should be able to drop the changes to this file.

anomiex commented on package.json:
```diff
@@ -108,119 +107,12 @@
-		"@wordpress/e2e-test-utils": "4.3.1",
-		"@wordpress/editor": "9.12.7",
-		"@wordpress/env": "1.6.0",
-		"@wordpress/eslint-plugin": "4.0.0",
```
@wordpress/eslint-plugin is still needed.
```$ yarn lint
yarn run v1.22.10
$ eslint --ext .js,.jsx . _inc extensions modules

Oops! Something went wrong! :(

ESLint: 7.14.0

ESLint couldn't find the plugin "@wordpress/eslint-plugin".
```

brbrr replied:
Ouch. Not sure why it was working for me. I guess there were some leftovers in `node_modules`

anomiex commented on package.json:
```diff
@@ -108,119 +107,12 @@
-		"css-loader": "5.0.1",
-		"danger": "10.5.4",
-		"enzyme": "3.11.0",
+		"babel-eslint": "10.1.0",
```
Alphabetize.

anomiex commented on package.json:
```diff
@@ -235,29 +127,8 @@
 		"eslint-plugin-wpcalypso": "5.0.0",
 		"glob": "7.1.6",
 		"husky": "4.3.0",
-		"jest": "26.6.3",
```
Since we're not installing jest anymore, we apparently need to declare the version in projects/plugins/jetpack/tests/e2e/.eslintrc.js (under settings.jest.version).
```
$ yarn lint
yarn run v1.22.10
$ eslint --ext .js,.jsx . _inc extensions modules

Oops! Something went wrong! :(

ESLint: 7.14.0

Error: Error while loading rule 'jest/no-deprecated-functions': Unable to detect Jest version - please ensure jest package is installed, or otherwise set version explicitly
Occurred while linting /usr/local/src/automattic/jetpack-reorg/projects/plugins/jetpack/tests/e2e/.eslintrc.js
```

anomiex reviewed (commented):
In addition to the inline comments, the definition for `yarn lint` needs updating to remove the extraneous `_inc extensions modules` directories.

brbrr committed:
Address feedback

brbrr committed:
Merge remote-tracking branch 'origin/feature/reorg' into fix/build-artifacts-action

anomiex committed:
Cleanup code coverage tests

anomiex committed:
Cleanup E2E test workflow

anomiex committed:
More fix attempts

anomiex committed:
More logging to try to debug

anomiex committed:
Do setup-node after setup-php

It looks like setup-php may be pushing /usr/local/bin onto the path,
which has a different version of node than the one setup-node is trying
to use.

anomiex committed:
Fix sectioning in .eslintignore

anomiex committed:
Readd dangerci

anomiex committed:
No need for dangerci in the plugin

anomiex committed:
Our dangerfile needs moment

kraftbj commented on .github/workflows/dangerci.yml:
```diff
@@ -10,7 +10,7 @@ jobs:
       - uses: actions/checkout@v2
 
-      - uses: actions/setup-node@v1
+      - uses: actions/setup-node@v2
```
We don't need setup-node anymore, we can use the .nvmrc file as la https://github.com/Automattic/jetpack/pull/18213/files#diff-3625957935f73eaca5408fdda40fce0a719a8551f975cefdc9249efa2cd840d3R22 , but we can pick that up in a follow-up PR since it doesn't negatively impact building.

anomiex replied:
At a quick glance through its documentation, it seems like nvm does more or less the same thing that setup-node does. It may be more flexible about things like asking for version "lts" instead of "12" or using non-official mirrors, but I don't think we need either of those things. On the down side, it seems it's per-shell, so I think we'd have to re-run it in every step that will be executing node (i.e. more boilerplate).

And setup-node@v2 may be better about taking advantage of

> The hosted images have been updated with the latest of each LTS from v8, v10, v12, and v14.

(quoted from https://github.com/actions/setup-node#v2)

kraftbj replied:
> On the down side, it seems it's per-shell, so I think we'd have to re-run it in every step that will be executing node (i.e. more boilerplate).

We should be able to set a default so it only has to be set at the start. The advantage in my mind is whenever we do bump to node 14, we only need to update .nvmrc and not all of the actions to match the version.

But in either case, we can dive into it in a future PR. No need to get sidetracked on it now.

kraftbj commented on package.json:
```diff
@@ -19,7 +19,6 @@
 		"build-client": "gulp",
 		"build-concurrently": "yarn install-if-deps-outdated && yarn clean && yarn concurrently 'yarn build-client' 'yarn build-php' 'yarn build-extensions' 'yarn build-search' 'yarn build-packages'",
 		"build-extensions": "webpack --config ./webpack.config.extensions.js",
-		"build-jetpack": "cd projects/plugins/jetpack && yarn build",
```
Not-blocking: What's the preferred way to build Jetpack if you're in the repo root with this removed?

anomiex replied:
FWIW, I added it back in #18212 along with cleaning up all the non-working things in this file.

kraftbj reviewed (changes requested):
We missed something. Installing this to JN drops Jetpack directly into the `plugins` dir (not in `plugins/jetpack-dev/` as before).

kraftbj commented on .github/workflows/build-artifacts.yml:
```diff
@@ -30,14 +30,14 @@ jobs:
       with:
         name: jetpack-dev
-        path: /tmp/artifact
+        path: /tmp/artifact/jetpack-dev/projects/plugins/jetpack
```
I think the difference is here. On L28, it builds into a `jetpack-dev` folder, but the path we are using here is directly into the plugin vs the level above it?

anomiex replied:
Hmm. Will the Automattic/action-jetpack-build-to-branch bit above also need updating somehow or other? ... Looks like yes, but only to change https://github.com/Automattic/action-jetpack-build-to-branch/blob/master/entrypoint.sh#L35 to use `tools/prepare-built-branch.sh` rather than `bin/prepare-built-branch.sh`. I have to wonder why we have that logic in a separate repo.

anomiex committed:
Fix artifact directory structure

We want everything inside a "jetpack-dev" folder.

kraftbj commented on .github/workflows/build-artifacts.yml:
```diff
@@ -22,22 +22,26 @@ jobs:
     - name: Inform beta builder
       run: |
-        VERSION=$(cat /tmp/artifact/jetpack-dev/version.txt)
+        VERSION=$(cat /tmp/artifact/jetpack-dev/projects/plugins/jetpack/version.txt)
```
```suggestion
        VERSION=$(cat /tmp/artifact/to-upload/jetpack-dev/version.txt)
```

anomiex replied:
D'oh.

anomiex committed:
Update .github/workflows/build-artifacts.yml

Co-authored-by: Brandon Kraft <public@brandonkraft.com>
